scripts/keymaker-create-account-for-iam-user

```diff
@@ -1,22 +1,55 @@
-#!/bin/bash -e
+#!/bin/bash
+
+# TODO: what to do if /etc/ssh/sshd_config isn't the running SSHD config file
+# TODO: more graceful handling of login banner (this is a stub) - libssh?
+# TODO: test user/group creation on: Amazon Linux, Suse, CentOS
 
 EX_TEMPFAIL=75
 EX_NOPERM=77
 
-if getent passwd "$PAM_USER" >/dev/null 2>&1; then
-    # Terminate the PAM authentication stack. The SSH client will fail since the user didn't supply a valid public key.
-    exit $EX_NOPERM
+PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
+SSHD_CONFIG="/etc/ssh/sshd_config"
+BANNER_FILE=`grep -e "^Banner." $SSHD_CONFIG | awk '{print $2}'`
+
+# Get the login banner
+if [ -e $BANNER_FILE ] ; then
+    BANNER=`cat $BANNER_FILE`
 else
-    # Create the user, then terminate the PAM authentication stack. The SSH client will fail, and the user will need to try again.
-    # TODO: figure out how to display info banner
-    # Verify that the IAM user exists.
-    keymaker get_authorized_keys "$PAM_USER" >/dev/null
+    BANNER=''
+fi
+
+create_account () {
     NEW_UID=$(keymaker get_uid "$PAM_USER")
     adduser "$PAM_USER" --disabled-password --gecos "$PAM_USER" --uid "$NEW_UID"
-    for group in $(keymaker get_groups "$PAM_USER"); do
-        usermod --append --groups "$group" "$PAM_USER" || echo "$0: Error while adding user to group"
+    groups=`keymaker get_groups "$PAM_USER"`
+    for group in $groups ; do
+        getent group | egrep "^$group:" >/dev/null
+        if [ "$?" == "1" ] ; then
+            echo "Creating group: $group"
+            groupadd $group
```

Comment:
There are two things going on.

First: As designed, Keymaker reconfigures sshd with `AuthenticationMethods publickey keyboard-interactive:pam,publickey`. The way this works is first `publickey` is tried, and only if that fails, then `keyboard-interactive:pam` AND `publickey` must succeed.

Keymaker also reconfigures pam so that `keyboard-interactive:pam` just runs `keymaker-create-account-for-iam-user` and then fails.

The fact that your sshd is asking you for a password means that this PAM configuration is not working; by design, it should just fail and short-circuit password authentication. Are you changing your sshd_config from what keymaker sets it to? Can you paste the contents of your sshd_config?

Second: You're wondering why you're only able to log in on the second try when provisioning an account. This is because when running PAM-based authentication, sshd second-guesses PAM by checking whether the username is present in the system - and it doesn't revisit this check after running the PAM stack. In other words, sshd will check for the username, run the PAM stack, then fail authentication anyway since the user account wasn't present when starting.

I'm looking for a way to get around this behavior and let the user log in on the first try, but so far haven't found anything.